alignborrowck: allow borrowing packed arrays with ABI align <= packed align in const generics

hns1971 opened the pull request:
Fixes #145376

### Summary
This PR fixes a false positive E0793 ("reference to packed field is unaligned")  
when borrowing from a `#[repr(C, packed)]` struct field whose type is a const-generic array  
with an element type that has ABI alignment <= the struct's packed alignment.

Example before this change:
```rust
#[repr(C, packed)]
struct PascalString<const CAP: usize> {
    len: u8,
    buf: [u8; CAP],
}

fn bar<const CAP: usize>(s: &PascalString<CAP>) -> &str {
    std::str::from_utf8(&s.buf[0..s.len as usize]).unwrap()
}

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16981989698/job/48143634856) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48143634856) [(plain)](https://github.com/rust-lang/rust/commit/99dedb01583e1ea116e010968028e4cf7f0c5056/checks/48143634856/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error: usage of qualified `ty::Ty<'tcx>`
  --> compiler/rustc_const_eval/src/util/alignment.rs:74:55
   |
74 | fn get_element_alignment<'tcx>(tcx: TyCtxt<'tcx>, ty: ty::Ty<'tcx>) -> Option<Align> {
   |                                                       ^^^^^^^^^^^^ help: try importing it and using it unqualified: `Ty<'tcx>`
   |
   = note: `-D rustc::usage-of-qualified-ty` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustc::usage_of_qualified_ty)]`

[RUSTC-TIMING] rustc_const_eval test:false 3.119
error: could not compile `rustc_const_eval` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] rustc_hir_analysis test:false 5.651
```

</details>

gralpli commented on compiler/rustc_const_eval/src/util/alignment.rs:
```diff
@@ -23,33 +23,79 @@ where
+/// Try to determine the alignment of an array element type
+fn get_element_alignment<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Align> {
+    match ty.kind() {
+        ty::Array(element_ty, _) => {
```
Can both branches be unified to ```ty::Array(element_ty, _) | ty::Slice(element_ty) => …```?

gralpli commented on tests/ui/packed/packed-array-const-generic.rs:
```diff
@@ -0,0 +1,19 @@
+}
+
+fn bar<const CAP: usize>(s: &PascalString<CAP>) -> &str {
+    // 目标：这行不应触发 E0793
```
I'm just the bug reporter, but I would guess that only English comments will be accepted into rustc, probably.

gralpli reviewed (commented):
Hey, I’m the bug reporter. Just what I noticed on a first glance.

lcnr commented on compiler/rustc_const_eval/src/util/alignment.rs:
```diff
@@ -23,33 +23,79 @@ where
+        Ok(normalized) => normalized,
+        Err(_) => {
+            // If normalization fails, fall back to the original type
+            ty
```
why? it feels a lot safer to conservatively return `true` here, same as when `layout_of` fails

lcnr commented on compiler/rustc_const_eval/src/util/alignment.rs:
```diff
@@ -23,33 +23,79 @@ where
-            true
+        Err(_) => {
+            // We cannot figure out the layout. This often happens with generic types.
+            // For const generic arrays like [u8; CAP], we can make a reasonable assumption
+            // about their alignment based on the element type.
```
we're not making a reasonable assumption. Their layout literally only depend son the layout of their element type, doesn't it?

hns1971 commented:
I've updated the PR as requested (renamed function, updated comment, removed special-casing). Just let me know if further adjustments are needed.

lcnr commented on compiler/rustc_const_eval/src/util/alignment.rs:
```diff
@@ -15,38 +15,74 @@ pub fn is_disaligned<'tcx, L>(
+fn get_element_alignment<'tcx>(_tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Align> {
+    match ty.kind() {
+        ty::Array(element_ty, _) | ty::Slice(element_ty) => {
+            // Only allow u8 and i8 arrays when layout computation fails
+            // Other types are conservatively assumed to be misaligned
+            match element_ty.kind() {
```
properly compute the of the `element_ty` here, limiting it to only u8 and i8 feels unnecessary :thinking:

lcnr commented:
Oh, yeah, thought this was a lint :+1: an FCP seems appropriate

hns1971 commented:
Thanks for the detailed feedback!

I’ve updated the PR with the following:
- Generalized the check to any `[T; N]`: when `layout_of([T; N])` is unavailable,
  we compute `layout_of(T)` and compare `abi_align(T)` to the packed alignment.
  This relies on `align([T]) == align(T)` (length is irrelevant).
  I also updated the comment with the proof sketch:
    (1) `&[T] -> &T` implies align([T]) >= align(T)
    (2) `std::array::from_ref(&T) -> &[T; 1]` implies align(T) >= align([T])
  Hence align([T]) == align(T).
- Renamed the helper to `is_potentially_misaligned` and switched terminology to “misaligned”.
  The doc comment now explicitly states that this may return false positives, but must not
  return false negatives (soundness-critical).
- Merged tests into a single file and renamed it to `packed-array-generic-length.rs`.
  The test covers:
    * allowed: `[u8; N]`, `[i8; N]`, `[NonZeroU8; N]`, `[MaybeUninit<u8>; N]`,
      and `#[repr(transparent)](u8)` wrappers
    * still error: `[u16; N]`, `[NonZeroU16; N]`, and `#[repr(transparent)](u16)` wrappers

I’ve also drafted the exact rules in the PR description and opened a Reference PR
documenting this behavior (see rust-lang/reference#1984).  
Please let me know if you want the rules tweaked further or if there’s a preferred place
in the Reference to put this.

tmandry commented:
Does this PR create a new semver hazard? For example..

```rust
// In a remote crate
struct Remote(u8);

// In local crate
#[repr(packed)]
struct Local {
  remote: Remote,
}

let remote = &Local::default().remote;
```

Then `Remote` can be changed to have a `u16`, which is now a semver-breaking change.

@rfcbot concern semver hazard

Otherwise, +1 for the overall idea.

@rfcbot reviewed

tmandry commented:
Hmm, I wish we could address the semver hazard with `repr(packed)`, but now that I understand what this PR actually does I can see it's not relevant.

@rfcbot resolve semver hazard

RalfJung commented:
> I do think that, from an implementation perspective, it's a bit icky for this particular check to do this dance of "fail to get layout, then do extra layout calculations". It feels like there ought to be a general query for "layout but I only care about alignment (or an under- or over-estimate thereof)" that everywhere that doesn't care about things like niches and field offsets, just the alignment, could use so that checks like this would always be consistent. (Like const prop of the Align UnOp in a MIR optimization arguably wants this as well.)

Yeah, agreed.

Seems worth at least filing a C-cleanup issue for?

hns1971 commented:
> > I do think that, from an implementation perspective, it's a bit icky for this particular check to do this dance of "fail to get layout, then do extra layout calculations". It feels like there ought to be a general query for "layout but I only care about alignment (or an under- or over-estimate thereof)" that everywhere that doesn't care about things like niches and field offsets, just the alignment, could use so that checks like this would always be consistent. (Like const prop of the Align UnOp in a MIR optimization arguably wants this as well.)
> 
> Yeah, agreed.
> 
> Seems worth at least filing a C-cleanup issue for?

Hi @RalfJung,

Thank you for the insightful architectural suggestion. I agree that a general `align_of` query would be a cleaner and more consistent solution, not only for this check but also for other parts of the compiler like const prop.

To keep the scope of this current PR focused and ensure it can be merged smoothly, I plan to **file a follow-up issue** to track this idea after this PR is merged. This will allow us to discuss and design the general query interface properly without blocking the immediate improvement here.

Thanks again for the review and the valuable idea.

RalfJung commented:
Sounds good. :) Note that I didn't even make the architectural suggestion though. Also, no need to be so formal in your writing :) . Short and to the point beats long and verbose.

bors commented:
:umbrella: The latest upstream changes (presumably #147145) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr commented on compiler/rustc_const_eval/src/util/alignment.rs:
```diff
@@ -15,38 +15,71 @@ pub fn is_disaligned<'tcx, L>(
+    match ty.kind() {
+        ty::Array(elem_ty, _) | ty::Slice(elem_ty) => {
+            // Try to obtain the element's layout; if we can, use its ABI align.
+            match tcx.layout_of(typing_env.as_query_input(*elem_ty)) {
```
^

bors commented:
:umbrella: The latest upstream changes (presumably #148721) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
